hack/build-integration-canary.sh

```diff
@@ -293,40 +293,55 @@ canary::build::integration(){
     docker_args+=(--build-arg "${shortsafename}_VERSION=$higher_readable")
   done
 
-
-  GO_VERSION="$(curl -fsSL "https://go.dev/dl/?mode=json&include=all" | jq -rc .[0].version)"
-  GO_VERSION="${GO_VERSION##*go}"
-  # If a release candidate, docker hub may not have the corresponding image yet.
-  # So, soften the version to just "rc", as they provide that as an alias to the latest available rc on their side
-  # See https://github.com/containerd/nerdctl/issues/3223
-  ! grep -Eq "rc[0-9]+$" <<<"$GO_VERSION" || GO_VERSION="${GO_VERSION%rc[0-9]*}-rc"
-  docker_args+=(--build-arg "GO_VERSION=$GO_VERSION")
+  docker_args+=(--build-arg "GO_VERSION=$(canary::golang::hublatest)")
 
   log::debug "${docker_args[*]} ."
   "${docker_args[@]}" "."
 }
 
+# Hub usually has a delay before available golang version show-up. This method will find the latest available one.
+# See
+# - https://github.com/containerd/nerdctl/issues/3224
+# - https://github.com/containerd/nerdctl/issues/3306
+canary::golang::hublatest(){
+  local hub_tags
+  local go_version
+  local available_version=""
+  local index
+
+  hub_tags="$(http::get /dev/stdout "https://registry-1.docker.io/v2/library/golang/tags/list" -H "Authorization: Bearer $(http::get /dev/stdout "https://auth.docker.io/token?service=registry.docker.io&scope=repository%3Alibrary%2Fgolang%3Apull" | jq  -rc .access_token)")"
+
+  index=0
+  while [ "$available_version" == "" ] && [ "$index" -lt 5 ]; do
```

Comment:
nit: why `5`? Don't we need to handle cases of `available_version` isn't found?

Reply:
Limiting the number of tries to 5.
If docker hub does not have any of the past five versions, all bets are off.
Should hard error in that case though.
Will do.

Reply:
Taking that back re-reading the code. It is fine to fail to find a version. In that case, the dockerfile will just fallback to the hard-coded version.

Reply:
Latest push has the tweak.
Thanks @ktock 
Appreciate the sharp review.

Reply:
And: hard error is not necessary (hub is notorious for random failures). The new check will allow the dockerfile to still build with the hardcoded version instead.